packages/contracts/test/helpers/SimpleWallet.sol

```diff
@@ -0,0 +1,141 @@
+    }
+    
+    function initialize() public initializer {
+        __Ownable_init();
+        isRecovering = false;
+    }
+
+    function transfer(address to, uint256 amount) public onlyNotRecoveringOwner {
+        require(address(this).balance >= amount, "insufficient balance");
+        payable(to).transfer(amount);
+    }
+
+    function withdraw(uint256 amount) public onlyNotRecoveringOwner {
+        transfer(msg.sender, amount);
+    }
+
+    function acceptanceSubjectTemplates()
+        public
+        pure
+        override
+        returns (string[][] memory) {
+        string[][] memory templates = new string[][](1);
+        templates[0] = new string[](5);
+        templates[0][0] = "Accept";
+        templates[0][1] = "guardian";
+        templates[0][2] = "request";
+        templates[0][3] = "for";
+        templates[0][4] = "{ethAddr}";
+        return templates;
+    }
+
+    function recoverySubjectTemplates()
+        public
+        pure
+        override
+        returns (string[][] memory) {
+        string[][] memory templates = new string[][](1);
+        templates[0] = new string[](8);
+        templates[0][0] = "Set";
+        templates[0][1] = "the";
+        templates[0][2] = "new";
+        templates[0][3] = "signer";
+        templates[0][4] = "of";
+        templates[0][5] = "{ethAddr}";
+        templates[0][6] = "to";
+        templates[0][7] = "{ethAddr}";
+        return templates;
+        }
+
+    function requestGuardian(address guardian) public onlyNotRecoveringOwner {
+        require(guardian != address(0), "invalid guardian");
+        require(guardians[guardian] == GuardianStatus.NONE, "invalid guardian status");
+        guardians[guardian] = GuardianStatus.REQUESTED;
+    }
+
+    function acceptGuardian(
+        address guardian,
+        uint templateIdx,
+        bytes[] memory subjectParams,
+        bytes32
+    ) internal override onlyNotRecoveringOwner {
+        require(guardian != address(0), "invalid guardian");
+        require(guardians[guardian] == GuardianStatus.REQUESTED, "invalid guardian status");
+        require(templateIdx == 0, "invalid template index");
+        require(subjectParams.length == 1, "invalid subject params");
+        (address guardianInEmail) = abi.decode(subjectParams[0], (address));
+        require(guardianInEmail == guardian, "invalid guardian in email");
+        guardians[guardian] = GuardianStatus.ACCEPTED;
+    }
+
+    function recoverWallet(
+        address guardian,
+        uint templateIdx,
+        bytes[] memory subjectParams,
+        bytes32
+    ) internal override onlyNotRecoveringOwner {
+        require(guardian != address(0), "invalid guardian");
+        require(guardians[guardian] == GuardianStatus.ACCEPTED, "invalid guardian status");
+        require(templateIdx == 0, "invalid template index");
+        require(subjectParams.length == 2, "invalid subject params");
+        // TODO Is the following implementation correct?
```

Comment:
Yes, we need subjectParams[1].
Sorry, my implementation was incorrect.
It will be
> (address walletAddrInEmail) = abi.decode(subjectParams[0], (address));
> (address newSigner) = abi.decode(subjectParams[1], (address));

Instead of `guardianInEmail == guardian`, it should check `walletAddrInEmail == address(this)`, i.e., the specified wallet address is the same as the called wallet address.

Reply:
@SoraSuegami I see, thank you for your information. I fixed the above points in feat/contracts branch.